cognite_toolkit/_cdf_tk/client/data_classes/capabilities.py

```diff
@@ -0,0 +1,170 @@
+            return ExtractionPipelineScope(ids=intersection_ids)
+        return None
+    elif isinstance(scope1, TableScope) and isinstance(scope2, TableScope):
+        intersection_tables = set(scope1.dbs_to_tables.keys()) & set(scope2.dbs_to_tables.keys())
+        if not intersection_tables:
+            return None
+        return TableScope(
+            dbs_to_tables={
+                db: sorted(set(scope1.dbs_to_tables[db]) & set(scope2.dbs_to_tables[db]))
+                for db in sorted(intersection_tables)
+            }
+        )
+    elif isinstance(scope1, InstancesScope) and isinstance(scope2, InstancesScope):
+        if intersection_instances := sorted(set(scope1.instances) & set(scope2.instances)):
+            return InstancesScope(instances=intersection_instances)
+        return None
+    elif isinstance(scope1, PartitionScope) and isinstance(scope2, PartitionScope):
+        if intersection_partition_ids := sorted(set(scope1.partition_ids) & set(scope2.partition_ids)):
+            return PartitionScope(partition_ids=intersection_partition_ids)
+        return None
+    elif isinstance(scope1, LegacySpaceScope) and isinstance(scope2, LegacySpaceScope):
+        if intersection_external_ids := sorted(set(scope1.external_ids) & set(scope2.external_ids)):
+            return LegacySpaceScope(external_ids=intersection_external_ids)
+        return None
+    elif isinstance(scope1, LegacyDataModelScope) and isinstance(scope2, LegacyDataModelScope):
+        if intersection_external_ids := sorted(set(scope1.external_ids) & set(scope2.external_ids)):
+            return LegacyDataModelScope(external_ids=intersection_external_ids)
+        return None
+    elif isinstance(scope1, ExperimentsScope) and isinstance(scope2, ExperimentsScope):
+        if intersection_experiments := sorted(set(scope1.experiments) & set(scope2.experiments)):
+            return ExperimentsScope(experiments=intersection_experiments)
+        return None
+    elif isinstance(scope1, PostgresGatewayUsersScope) and isinstance(scope2, PostgresGatewayUsersScope):
+        if intersection_usernames := sorted(set(scope1.usernames) & set(scope2.usernames)):
+            return PostgresGatewayUsersScope(usernames=intersection_usernames)
+        return None
+    else:
+        raise TypeError(f"Unknown scope type {type(scope1)!r}")
+
+
+def scope_union(scope1: T_Scope, scope2: T_Scope | None) -> Capability.Scope:
+    """
+    Returns the union of two scopes.
+    If both scopes are None, returns None.
+    If one scope is None, returns the other scope.
+    If both scopes are the same, returns the union of the two scopes.
+    If scopes are different types, raise an ValueError.
+    """
+    if scope2 is None:
+        return scope1
+    if type(scope1) is not type(scope2):
+        raise ValueError("Cannot union scopes of different types")
+    if isinstance(scope1, AllScope) or isinstance(scope2, AllScope):
+        return AllScope()
```

Comment:
This code is duplicated. scope_intersection has the same in the initial part of the if/else chain. Can you refactor that to avoid duplication?